File: flutter/lib/src/models/action.dart
import 'enums.dart';

sealed class NDAction {
  const NDAction();

  factory NDAction.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String? ?? '';
    return switch (type) {
      'open_url' => OpenUrlAction.fromJson(json),
      'custom' => CustomAction.fromJson(json),
      'navigate' => NavigateAction.fromJson(json),
      'event' => TrackEventAction.fromJson(json),
      'composite' => CompositeAction.fromJson(json),
      _ => CustomAction(key: type, value: null),
    };
  }
}

class OpenUrlAction extends NDAction {
  // url can be a plain string, or a platform map {"android": "...", "ios": "..."}
  final String url;
  final bool openInBrowser;
  final bool customTabsEnabled;

  const OpenUrlAction({
    required this.url,
    this.openInBrowser = false,
    this.customTabsEnabled = true,
  });

  factory OpenUrlAction.fromJson(Map<String, dynamic> json) {
    final urlRaw = json['url'];
    String url;
    if (urlRaw is Map) {
      // Platform-specific URL map — use 'ios' key for Flutter (treated as iOS-like)
      url = (urlRaw['ios'] as String?) ?? (urlRaw['android'] as String?) ?? '';
    } else {
      url = urlRaw as String? ?? '';
    }
    return OpenUrlAction(
      url: url,
      openInBrowser: json['openInBrowser'] as bool? ?? false,
      customTabsEnabled: json['customTabsEnabled'] as bool? ?? true,
    );
  }
}

class CustomAction extends NDAction {
  final String key;
  final dynamic value;
  final Map<String, String>? metadata;

  const CustomAction({required this.key, this.value, this.metadata});

  factory CustomAction.fromJson(Map<String, dynamic> json) => CustomAction(
        key: json['key'] as String? ?? '',
        value: json['value'],
        metadata: (json['metadata'] as Map<String, dynamic>?)
            ?.map((k, v) => MapEntry(k, v as String)),
      );
}

class NavigateAction extends NDAction {
  final String destination;
  final Map<String, String>? params;

  const NavigateAction({required this.destination, this.params});

  factory NavigateAction.fromJson(Map<String, dynamic> json) => NavigateAction(
        destination: json['destination'] as String? ?? '',
        params: (json['params'] as Map<String, dynamic>?)
            ?.map((k, v) => MapEntry(k, v as String)),
      );
}

class TrackEventAction extends NDAction {
  final String eventName;
  final Map<String, dynamic>? properties;

  const TrackEventAction({required this.eventName, this.properties});

  factory TrackEventAction.fromJson(Map<String, dynamic> json) => TrackEventAction(
        eventName: json['eventName'] as String? ?? '',
        properties: json['properties'] as Map<String, dynamic>?,
      );
}

class CompositeAction extends NDAction {
  final List<NDAction> actions;
  final ExecutionMode executionMode;

  const CompositeAction({
    required this.actions,
    this.executionMode = ExecutionMode.sequential,
  });

  factory CompositeAction.fromJson(Map<String, dynamic> json) => CompositeAction(
        actions: (json['actions'] as List?)
                ?.map((e) => NDAction.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        executionMode: ExecutionMode.fromJson(json['executionMode'] as String? ?? 'sequential'),
      );
}

abstract final class ActionTriggers {
  static const onClick = 'onClick';
  static const onLongPress = 'onLongPress';
  static const onDoubleTap = 'onDoubleTap';
  static const onAppear = 'onAppear';
  static const onDisappear = 'onDisappear';
}

File: react-native/src/models/Action.ts
import { ExecutionMode } from './enums';

/**
 * The shape of a per-platform URL entry. Production payloads ship two
 * variants:
 *
 *  1. A flat string: `{ "android": "https://...", "ios": "https://..." }`.
 *  2. A nested object carrying `text` (+ optional `replacements`), the legacy
 *     Ultron shape: `{ "android": { "text": "https://...", "replacements": {} } }`.
 *
 * Either is accepted; `resolveOpenUrl` flattens the nested shape down to a
 * string. `replacements` are not expanded yet on RN (Android/iOS also just
 * read `text` today).
 */
export type PlatformUrlValue =
  | string
  | { text?: string; replacements?: Record<string, string> };

export interface OpenUrlAction {
  type: 'open_url';
  url:
    | string
    | { android?: PlatformUrlValue; ios?: PlatformUrlValue; rn?: PlatformUrlValue };
  openInBrowser?: boolean;
  customTabsEnabled?: boolean;
  /**
   * Server-injected attribution fields (`wzrk_element_id`, `wzrk_c2a`, etc.).
   * Flattened into the click event's `additionalProperties` via
   * `ActionAttributionExtras`.
   */
  metadata?: Record<string, string>;
}

export interface CustomAction {
  type: 'custom';
  key: string;
  value?: unknown;
  metadata?: Record<string, string>;
}

export interface NavigateAction {
  type: 'navigate';
  destination: string;
  params?: Record<string, string>;
}

export interface TrackEventAction {
  type: 'event';
  eventName: string;
  properties?: Record<string, unknown>;
}

export interface CompositeAction {
  type: 'composite';
  actions: Action[];
  executionMode: ExecutionMode;
}

export type Action =
  | OpenUrlAction
  | CustomAction
  | NavigateAction
  | TrackEventAction
  | CompositeAction;

export type ActionTrigger =
  | 'onClick'
  | 'onLongPress'
  | 'onDoubleTap'
  | 'onAppear'
  | 'onDisappear';

/**
 * Pick the platform-correct URL from an `OpenUrlAction.url` field.
 *
 * Preference order is `rn` → `android` → `ios`. Each entry can be either a
 * plain string or the legacy nested `{ text, replacements }` shape; the
 * nested form is flattened to its `text` value.
 */
export function resolveOpenUrl(action: OpenUrlAction): string {
  if (typeof action.url === 'string') return action.url;
  return (
    flattenPlatformUrl(action.url.rn) ||
    flattenPlatformUrl(action.url.android) ||
    flattenPlatformUrl(action.url.ios) ||
    ''
  );
}

function flattenPlatformUrl(value: PlatformUrlValue | undefined): string {
  if (value == null) return '';
  if (typeof value === 'string') return value;
  return value.text ?? '';
}
